src/protocol/transport_service.rs

```diff
@@ -96,6 +96,167 @@ impl ConnectionContext {
+/// the timeout is reset.
+#[derive(Debug)]
+struct KeepAliveTracker {
+    /// Close the connection if no substreams are open within this time frame.
+    keep_alive_timeout: Duration,
+
+    /// Track substream last activity.
+    last_activity: HashMap<(PeerId, ConnectionId), Instant>,
+
+    /// Pending keep-alive timeouts.
+    pending_keep_alive_timeouts: FuturesUnordered<BoxFuture<'static, (PeerId, ConnectionId)>>,
+
+    /// Saved waker.
+    waker: Option<Waker>,
+}
+
+impl KeepAliveTracker {
+    /// Create new [`KeepAliveTracker`].
+    pub fn new(keep_alive_timeout: Duration) -> Self {
+        Self {
+            keep_alive_timeout,
+            last_activity: HashMap::new(),
+            pending_keep_alive_timeouts: FuturesUnordered::new(),
+            waker: None,
+        }
+    }
+
+    /// Called on connection established event to add a new keep-alive timeout.
+    pub fn on_connection_established(&mut self, peer: PeerId, connection_id: ConnectionId) {
+        self.substream_activity(peer, connection_id);
+    }
+
+    /// Called on connection closed event.
+    pub fn on_connection_closed(&mut self, peer: PeerId, connection_id: ConnectionId) {
+        self.last_activity.remove(&(peer, connection_id));
+    }
+
+    /// Called on substream opened event to track the last activity.
+    pub fn substream_activity(&mut self, peer: PeerId, connection_id: ConnectionId) {
+        // Keep track of the connection ID and the time the substream was opened.
+        if self.last_activity.insert((peer, connection_id), Instant::now()).is_none() {
+            // Refill futures if there is no pending keep-alive timeout.
+            let timeout = self.keep_alive_timeout;
+            self.pending_keep_alive_timeouts.push(Box::pin(async move {
+                tokio::time::sleep(timeout).await;
+                (peer, connection_id)
+            }));
+        }
+
+        tracing::trace!(
+            target: LOG_TARGET,
+            ?peer,
+            ?connection_id,
+            ?self.keep_alive_timeout,
+            last_activity = ?self.last_activity,
+            pending_keep_alive_timeouts = ?self.pending_keep_alive_timeouts.len(),
+            "substream activity",
+        );
+
+        // Wake any pending poll.
+        self.waker.take().map(|waker| waker.wake());
+    }
+}
+
+impl Stream for KeepAliveTracker {
+    type Item = (PeerId, ConnectionId);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.pending_keep_alive_timeouts.is_empty() {
+            // No pending keep-alive timeouts.
+            self.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        match self.pending_keep_alive_timeouts.poll_next_unpin(cx) {
+            Poll::Ready(Some(key)) => {
+                // Check last-activity time.
+                let Some(when) = self.last_activity.get(&key) else {
```

Comment:
Nit for easier understanding of what time point `when` is referring to:
```suggestion
                let Some(last_activity) = self.last_activity.get(&key) else {
```